### ISSUE TITLE
Logging & Error Handling

### ISSUE THREAD

**Issue by splitbrain** [splitbrain/dokuwiki#3354]
Earlier today I merged #3203 and #3230 into master, introducing central error handling, logging and a log viewer.

I would like people to give it a try and provide feedback on the usefulness of the current implementation. I would especially like to hear thoughts on:

* The lack of log rotation/deletion. Should DokuWiki try to clean up old logs?
* The lack of error levels and PSR-3 compatibility. Personally I believe the facility based approach is simpler and more suited to DokuWiki.
* The current log format
* The lack of plugin events during logging
* The log viewer interface
* The mechanism to enable/disable facilities and how this applies to new facilities

**Comment by mprins** [comment]
DW should rotate or cleanup or overwrite logs it generates as setting up a logrotation with external tooling is basically impossible a in shared hosting environment.

**Comment by iainhallam** [comment]
@splitbrain On the lack of error levels, I'm starting to write a plugin which will do a fair amount of logging. I'm wondering which is best - to create a new facility that has all the logs for the plugin, or only to log notice/info level things there and to put debug into the debug facility and errors into the error facility. Each approach has pros and cons - the ability to find everything for the plugin in one place vs. the ability to find all errors in one place, mainly.

Personally, I would find error levels quite useful, as in many places, facilities are tied to the subsystem that generates the message, rather than levels. I could place all my plugin logs into one facility, and the user could choose which levels to see in the viewer. It would also be nice to have the viewer able to see all the facilities in one stream (picking the level separately), or to see them separately by facility as it does at the moment.

Ultimately, it seems like facilities currently try to do two jobs, which will make it difficult for plugin authors to know what the best approach is, and you'll likely have different people doing different things, which then makes life difficult for the admin of a site to know where to find things.

TL;DR: facilities for subsystems, including DokuWiki itself and plugins, with error levels as an orthogonal concept, would be my preference.

**Comment by iainhallam** [comment]
Out of interest, what does "The lack of plugin events during logging" mean in practice? I've yet to jump into the code to look through the logging system, and this doesn't seem to me to appear in the docs that have surfaced so far.

**Comment by splitbrain** [comment]
@iainhallam that kind of feedback would have been really helpful. like one and half years ago ;-)

**Comment by iainhallam** [comment]
> @iainhallam that kind of feedback would have been really helpful. like one and half years ago ;-)

I'm sorry - have only just been looking at the system as I've had very little time since the start of the pandemic. We can iterate, though, right? Is there a way to see how many plugins are already using the system?

**Comment by Klap-in** [comment]
All error, debug and deprecated messages are logged, regardless there origin. So also these from plugins. There are some older, now deprecated, functions replaced. These can still be expected in code of plugins.

further you could use codesearch for tracing adoption of the new functions 
[codesearch Logger::error(](https://codesearch.dokuwiki.org/search?full=%22Logger%3A%3Aerror%5C%28%22&defs=&refs=&path=&hist=&type=&xrd=&nn=3&si=full&searchall=true&si=full)
[codesearch Logger::debug(](https://codesearch.dokuwiki.org/search?full=%22Logger%3A%3Adebug%5C%28%22&defs=&refs=&path=&hist=&type=&xrd=&nn=3&si=full&searchall=true&si=full)
[codesearch Logger::deprecated(](https://codesearch.dokuwiki.org/search?full=%22Logger%3A%3Adeprecated%5C%28%22&defs=&refs=&path=&hist=&type=&xrd=&nn=3&si=full&searchall=true&si=full)

or the old one
[codesearch dbglog](https://codesearch.dokuwiki.org/search?project=dokuwiki&project=plugin&project=template&full=dbglog)

**Comment by iainhallam** [comment]
It looks as though codesearch just covers the bundled plugins. I was wondering if there's anything that looks across all plugins registered at dokuwiki.org, like the previous plugin surveys look to have done, or whether those were run manually.

**Comment by Klap-in** [comment]
The codesearch is the best tool at the moment because it searches in the latest version of the source code at the download url known at the plugins wiki page. Of course what you find depends on getting the precise search terms. 

In this case indeed only bundled plugins has implemented the new functionality. It is just released. The old functionality is still working, while the new functionality is just available since the last version was released. 

Normally, in DokuWiki source code we deprecated functionality in two releases. So, this release we mark it as deprecated, the next it can be deleted. However, sometimes the moment of deleting is delayed if it is used too much and/or rewriting is a lot of work. (Some big changes are actively brought to attention of the extension authors in their issue tracker.)

For a lot of plugins is having overlap with current and last version practical, and therefore these replace it a bit later at a point where you know a lot of people have the newest version, so less people will be affected because they update first their wiki and next their plugins.

To help plugin/template authors we have next to the detailed developers’ changelog (https://www.dokuwiki.org/devel:releases), also guidelines for updating of deprecated code:

- https://www.dokuwiki.org/devel:releases:refactor2021 (btw: above suggestions about new debug functions I still like to add)
- https://www.dokuwiki.org/devel:releases:refactor2020

(Maybe we can share these guides bit more active in the newsletter, if people consider that is appropriate).
Combined with logging of the deprecated code accessible via the new Logviewer, this could assist plugin authors or people who like to help them with tracing recent changes.

In the guidelines you can find a lot of the codesearchs to old code. And if you need examples, codesearch for the new functionality can give you often helpful examples of current implementations.

The plugin survey was very nice, I like it as well (I have even considered to continue it in certain way), but it is also a lot of work. Therefore, instead codesearch, the guidelines above and better accessible logging are introduced.

@splitbrain is also busy with updating of xref (https://github.com/splitbrain/dokuwiki-plugin-xref/pull/5) Another helpful tool.
